plugins/account_history_plugin/account_history_plugin.cpp

```diff
@@ -48,45 +70,190 @@ optional<block_id_type> account_history_plugin_impl::find_block_id(const transac
-      {
-         for (const auto& cycle : block->cycles)
-            for (const auto& thread : cycle)
-               for (const auto& trx : thread.user_input)
-                  if (trx.id() == transaction_id)
-                     return trx;
-      }
-
-      // ERROR in indexing logic
-      FC_ASSERT(block, "Transaction with ID ${tid} was indexed as being in block ID ${bid}, but no such block was found", ("tid", transaction_id)("bid", block_id));
-      FC_THROW("Transaction with ID ${tid} was indexed as being in block ID ${bid}, but was not found in that block", ("tid", transaction_id)("bid", block_id));
+      return find_transaction(transaction_id, *block_id);
    }
 
 #warning TODO: lookup of recent transactions
    FC_THROW_EXCEPTION(chain::unknown_transaction_exception,
                       "Could not find transaction for: ${id}", ("id", transaction_id.str()));
 }
 
+get_transactions_results account_history_plugin_impl::get_transactions(const AccountName&  account_name, const optional<uint32_t>& start_seq, const optional<uint32_t>& stop_seq) const
+{
+   fc::time_point start_time = fc::time_point::now();
+   const auto& db = chain_plug->chain().get_database();
+
+   block_transaction_id_map block_transaction_ids;
+   db.with_read_lock( [&]() {
+      const auto& account_idx = db.get_index<transaction_history_multi_index, by_account_name>();
+      auto range = account_idx.equal_range( account_name );
+      for (auto transaction_history = range.first; transaction_history != range.second; ++transaction_history)
+      {
+         block_transaction_ids.emplace(std::make_pair(transaction_history->block_id, transaction_history->transaction_id));
+      }
+   } );
+
+   uint32_t begin, end;
+   const auto size = block_transaction_ids.size();
+   if (!start_seq)
+   {
+      begin = 0;
+      end = size;
+   }
+   else
+   {
+      begin = *start_seq;
+
+      if (!stop_seq)
+         end = size;
+      else
+      {
+         end = *stop_seq + 1;
+         if (end > size)
+            end = size;
+      }
+
+      if (begin > size - 1 || begin >= end)
+         return get_transactions_results();
+   }
+
+   return ordered_transactions(block_transaction_ids, start_time, begin, end);
+}
+
+get_transactions_results account_history_plugin_impl::ordered_transactions(const block_transaction_id_map& block_transaction_ids, const fc::time_point& start_time, const uint32_t begin, const uint32_t end) const
+{
+   get_transactions_results results;
+   results.transactions.reserve(end - begin);
+
+   const auto& db = chain_plug->chain().get_database();
+   auto block_transaction = block_transaction_ids.cbegin();
+
+   uint32_t current = 0;
+   // keep iterating through each equal range
+   while (block_transaction != block_transaction_ids.cend() && !time_exceeded(start_time))
+   {
+      optional<signed_block> block;
+      db.with_read_lock( [&]() {
+         block = chain_plug->chain().fetch_block_by_id(block_transaction->first);
+      } );
```

Comment:
`fetch_block_by_id` doesn't need a lock, since it's not hitting the database at all. Blocks aren't stored in the database. We don't want those in RAM.